## 一、心血管疾病风险预测

### Step 1：数据分析

In [1]:
ls '/home/kesci/input/diabetes8566/'

hw_Cardiovascular.csv  hw_Nephropathy.csv  test.csv  test_y.csv  train.csv


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
# 加载数据
card_df = pd.read_csv('/home/kesci/input/diabetes8566/hw_Cardiovascular.csv',encoding='utf-8')
card_df.head()

Age  Sex  Body.height..at.0M  Waist.circumstances.at.0M  Duration.of.T2DM  \
0   61    1               170.0                    101.000              72.0   
1   42    1               184.0                     99.532              36.0   
2   64    1               169.0                     94.627             168.0   
3   57    1               168.5                     96.122             144.0   
4   63    2               151.0                     88.500              72.0   

   Smoking.habit  Drinking.habit  History.of.complications  \
0              0               0                         1   
1              2               0                         1   
2              2               0                         1   
3              2               1                         1   
4              1               0                         1   

   History.of.hypertension  History.of.dyslipidemia  ...  Urinary.sugar.at.3M  \
0                        1                        1  ...                    0   
1                        0                        1  ...                    4   
2                        0                        1  ...                    4   
3                        1                        1  ...                    0   
4                        1                        1  ...                    0   

   Urinary.sugar.at.12M  Urinary.albumin.creatinine.ratio.at.0M  \
0                     0                                  53.300   
1                     3                                  84.301   
2                     0                                  14.497   
3                     0                                  19.916   
4                     0                                  30.000   

   Urinary.albumin.creatinine.ratio.at.3M  \
0                                  94.218   
1                                  76.349   
2                                  13.418   
3                                  19.455   
4                                  27.100   

   Urinary.albumin.creatinine.ratio.at.12M  eGFR.at.0M  eGFR.at.1M  \
0                                    6.400   55.940643   52.670513   
1                                   58.541   98.878513   96.060985   
2                                   12.110   83.486588   83.701169   
3                                   16.512   85.148398   85.273085   
4                                   56.600   77.754106   87.433453   

   eGFR.at.3M  eGFR.at.12M  Cardiovascular Risk  
0   50.686818    61.643003                    1  
1   96.013865    89.393232                    0  
2   80.666781    79.474729                    0  
3   83.015438    79.775872                    1  
4   89.274571    83.961202                    0  

[5 rows x 165 columns]

In [8]:
print(card_df.shape)  #数据量大小
round(card_df.describe()) 

(778, 165)


Age    Sex  Body.height..at.0M  Waist.circumstances.at.0M  \
count  778.0  778.0               778.0                      778.0   
mean    64.0    1.0               163.0                       88.0   
std     12.0    0.0                10.0                       10.0   
min     27.0    1.0               135.0                       38.0   
25%     56.0    1.0               155.0                       82.0   
50%     64.0    1.0               164.0                       88.0   
75%     73.0    2.0               170.0                       93.0   
max     98.0    2.0               188.0                      187.0   

       Duration.of.T2DM  Smoking.habit  Drinking.habit  \
count             778.0          778.0           778.0   
mean              106.0            1.0             0.0   
std                78.0            1.0             1.0   
min                 1.0            0.0             0.0   
25%                48.0            0.0             0.0   
50%                96.0            0.0             0.0   
75%               135.0            2.0             1.0   
max               480.0            2.0             1.0   

       History.of.complications  History.of.hypertension  \
count                     778.0                    778.0   
mean                        1.0                      1.0   
std                         0.0                      0.0   
min                         0.0                      0.0   
25%                         1.0                      0.0   
50%                         1.0                      1.0   
75%                         1.0                      1.0   
max                         1.0                      1.0   

       History.of.dyslipidemia  ...  Urinary.sugar.at.3M  \
count                    778.0  ...                778.0   
mean                       1.0  ...                  1.0   
std                        0.0  ...                  1.0   
min                        0.0  ...                  0.0   
25%                        0.0  ...                  0.0   
50%                        1.0  ...                  0.0   
75%                        1.0  ...                  0.0   
max                        1.0  ...                  5.0   

       Urinary.sugar.at.12M  Urinary.albumin.creatinine.ratio.at.0M  \
count                 778.0                                   778.0   
mean                    1.0                                    85.0   
std                     1.0                                   217.0   
min                     0.0                                     0.0   
25%                     0.0                                    11.0   
50%                     0.0                                    24.0   
75%                     0.0                                    64.0   
max                     5.0                                  3451.0   

       Urinary.albumin.creatinine.ratio.at.3M  \
count                                   778.0   
mean                                     80.0   
std                                     192.0   
min                                       0.0   
25%                                      13.0   
50%                                      30.0   
75%                                      67.0   
max                                    3525.0   

       Urinary.albumin.creatinine.ratio.at.12M  eGFR.at.0M  eGFR.at.1M  \
count                                    778.0       778.0       778.0   
mean                                      97.0        79.0        77.0   
std                                      274.0        19.0        19.0   
min                                        0.0        17.0        16.0   
25%                                       11.0        67.0        65.0   
50%                                       26.0        78.0        76.0   
75%                                       71.0        90.0        89.0   
max                                     3678.0       155.0       158.0   

       eGFR.at.3M  eGFR.

In [10]:
# 查看是否有缺失值
print(card_df.isnull().sum()) #判断每列缺失值的数量

Age                                        0
Sex                                        0
Body.height..at.0M                         0
Waist.circumstances.at.0M                  0
Duration.of.T2DM                           0
Smoking.habit                              0
Drinking.habit                             0
History.of.complications                   0
History.of.hypertension                    0
History.of.dyslipidemia                    0
History.of.hyperuricemia                   0
History.of.retinopathy                     0
History.of.arteriosclerosis.obliterans     0
History.of.atrial.fibrillation             0
History.of.kidney.disease                  0
History.of.liver.disease                   0
History.of.myocardial.infarction           0
History.of.cerebral.infarction             0
History.of.angina.pectoris                 0
History.of.heart.failure                   0
Systolic.blood.pressure.at.0M              0
Systolic.blood.pressure.at.1M              0
Systolic.b

In [13]:
# 查看数据类别是否平衡
fig, ax = plt.subplots()
sns.set()
ax = sns.distplot(card_df['Cardiovascular Risk'],kde=False)

print(card_df['Cardiovascular Risk'].value_counts())

0    703
1     75
Name: Cardiovascular Risk, dtype: int64


<Figure size 432x288 with 1 Axes>

**分析结果**：由上面的基本分析，该数据包含778个样本，165列，其中最后一列Cardiovascular Risk为label。由label的值可知该问题**是个二分类问题**。
数据集中不存在缺失值。但是数据**存在极大的类别不平衡**，标签为0的有703个样本，而标签为1的只有75个。

### Step 2：数据预处理

1. 训练，测试和标签数据划分；
1. 数据归一化：由于不同特征之间的取值范围不一样，进行归一化可加速模型收敛；参考：https://www.jianshu.com/p/95a8f035c86c
1. 类别不平衡处理：由于该数据集中正样本数量较少，而在实际临床中，**正样本（发送心血管疾病）的预测远远比一个负样本的预测重要**。因此可采用加权处理，过采样或降采样的方法。参考：https://mp.weixin.qq.com/s/pOJof9aGp8iRlJuKH_Iz-A?

In [6]:
#把标签列删除
card_X = card_df.drop(columns='Cardiovascular Risk')
#把标签列取出
card_Y = card_df['Cardiovascular Risk']
card_Y.head()

0    1
1    0
2    0
3    1
4    0
Name: Cardiovascular Risk, dtype: int64

In [7]:
from sklearn import preprocessing

min_max_scaler = preprocessing.MinMaxScaler() 
card_X = min_max_scaler.fit_transform(card_X)
print(card_X)

[[0.47887324 0.         0.65779468 ... 0.25692934 0.25408388 0.38730928]
 [0.21126761 0.         0.92395437 ... 0.56298794 0.58138568 0.60206425]
 [0.52112676 0.         0.63878327 ... 0.47580686 0.47056601 0.5253064 ]
 ...
 [0.71830986 1.         0.19201521 ... 0.50364594 0.49439476 0.57402128]
 [0.22535211 0.         0.7338403  ... 0.55142926 0.53073369 0.63003969]
 [0.64788732 0.         0.58174905 ... 0.37515457 0.39613001 0.46211443]]


In [11]:
from sklearn.model_selection import train_test_split

#进行数据集划分，70%作为训练， 30作为测试
cx_train,cx_test,cy_train,cy_test = train_test_split(card_X, card_Y, test_size=0.3)

print(cx_train.shape,cy_train.shape)
print(cx_test.shape,cy_test.shape)

(544, 164) (544,)
(234, 164) (234,)


### Step 3:  模型构建：模型选择，训练
作为机器学习或统计学习模型，构建过程最基本的要考虑训练集与验证集，建模过程要通过交叉检验来完成，这里我们使用sklearn 包提供的机器学习框架。

对于类别不平衡，我们采用在模型训练过程中，通过设置损失函数的权重进行加权处理，并对比处理后和不处理的效果。

In [15]:
import xgboost as xgboost
from sklearn.ensemble import RandomForestClassifier
import sklearn.svm as svm
from sklearn.model_selection import KFold, cross_val_score

#10折交叉验证
kfold = KFold(n_splits=10, random_state=6) #10折交叉验证

#随机森林
forest = RandomForestClassifier()
rt = cross_val_score(forest, card_X, card_Y, cv=kfold)
print('随机森林 Accuracy：',rt.mean())

#SVM
sv_k = svm.SVC()
sv = cross_val_score(sv_k, card_X, card_Y, cv=kfold)
print('SVM Accuracy：',sv.mean())

#xgboost
xgboost = xgboost.XGBClassifier()
xgb = cross_val_score(xgboost, card_X, card_Y, cv=kfold)
print('xgboost Accuracy：',xgb.mean())

/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The de

随机森林 Accuracy： 0.9009990009990009


/opt/conda/lib/python3.6/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/opt/conda/lib/python3.6/site-packages/sklearn/svm/base.py:193: FutureWarning: The default 

SVM Accuracy： 0.9035797535797536
xgboost Accuracy： 0.9126040626040627


经过10折交叉验证，发现xgboost 模型（0.913）性能比随机森林（0.901）和SVM（0.903）略好，因此我们选择xgboost进行训练，测试

In [97]:
import xgboost as xgb
# xgboot： https://www.cnblogs.com/pinard/p/11114748.html 
#训练模型: 未处理类别不平衡问题
xgboost = xgb.XGBClassifier(learning_rate=0.01, max_depth=10, gamma=0.2) 
xgboost.fit(cx_train, cy_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bytree=1, gamma=0.2, learning_rate=0.01,
              max_delta_step=0, max_depth=10, min_child_weight=1, missing=None,
              n_estimators=100, n_jobs=1, nthread=None,
              objective='binary:logistic', random_state=0, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=1, seed=None, silent=True,
              subsample=1)

In [98]:
from sklearn.metrics import accuracy_score,precision_score,recall_score,confusion_matrix
from sklearn.metrics import roc_auc_score
#测试模型
cy_pred = xgboost.predict(cx_test)

print("Acc: " ,accuracy_score(cy_pred, cy_test))

print(confusion_matrix(cy_pred, cy_test))

Acc:  0.9273504273504274
[[212  17]
 [  0   5]]


In [99]:
# 类别平衡：https://www.cnblogs.com/qi-yuan-008/p/11992156.html
# 使用scale_pos_weight方法对lable不平衡问题进行处理 https://machinelearningmastery.com/xgboost-for-imbalanced-classification/ 
from sklearn.utils.class_weight import compute_sample_weight
from collections import Counter
cnt = Counter(cy_train)
xgboost2 = xgb.XGBClassifier(max_depth=10, gamma=0.2, scale_pos_weight=cnt[0]/cnt[1])  #设置类别平衡
xgboost2.fit(cx_train, cy_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bytree=1, gamma=0.2, learning_rate=0.1,
              max_delta_step=0, max_depth=10, min_child_weight=1, missing=None,
              n_estimators=100, n_jobs=1, nthread=None,
              objective='binary:logistic', random_state=0, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=9.264150943396226, seed=None,
              silent=True, subsample=1)

In [100]:
#测试模型
#测试模型
cy_pred2 = xgboost2.predict(cx_test)

print("Acc: " ,accuracy_score(cy_pred2, cy_test))

print(confusion_matrix(cy_pred2, cy_test))

Acc:  0.8974358974358975
[[204  16]
 [  8   6]]


**ROC曲线绘制**

In [101]:
from sklearn.metrics import roc_curve
cy_pred_p = xgboost.predict_proba(cx_test)
cy_pred_p2 = xgboost2.predict_proba(cx_test)
fpr, tpr, thresholds = roc_curve(cy_test, cy_pred_p[:,1], pos_label=1)
fpr2, tpr2, thresholds2 = roc_curve(cy_test,cy_pred_p2[:,1], pos_label=1)

AUC_ROC = roc_auc_score(cy_test, cy_pred_p[:,1])
AUC_ROC2 = roc_auc_score(cy_test, cy_pred_p2[:,1])

roc_curve =plt.figure()
plt.plot(fpr,tpr,'-',label='xgboost (AUC = %0.4f)' % AUC_ROC)
plt.plot(fpr2,tpr2,'-',label='xgboost2 (AUC = %0.4f)' % AUC_ROC2)
plt.title('ROC curve')
plt.xlabel("FPR (False Positive Rate)")
plt.ylabel("TPR (True Positive Rate)")
plt.legend(loc="lower right")

<Figure size 432x288 with 1 Axes>

### Step 4：总结分析

从上面可知，对于心血管数据集，使用Xgboost模型进行分类，可以达到比较好的分类准确率。
同时，我们对数据类别不平衡进行了研究处理，通过设置xgboost的scale_pos_weight参数进行加权后，虽然模型的准确率有所下降，但是对于二分类问题，以及这种不平衡的数据，通过比较正样本的AUC值具有比较大的意义，从上面的实验可得，**对样本类别进行加权处理后，模型在测试集的AUC值略有提升**。

不过，对于实际临床来数，以上两个模型在心血管疾病正样本的预测精确率上还有很大的提升空间。可继续深入研究

## 二、肾病风险预测

### Step 1：数据分析

In [3]:
# 加载数据
neph_df = pd.read_csv('/home/kesci/input/diabetes8566/hw_Nephropathy.csv',encoding='utf-8')
neph_df.head()

Age  Sex  Body height  at 0M   Waist circumstances at 0M   \
0  61.0  1.0                170.0                       101.0   
1  42.0  1.0                184.0                         NaN   
2  64.0  1.0                169.0                         NaN   
3  57.0  1.0                168.5                         NaN   
4  63.0  2.0                151.0                        88.5   

   Duration of T2DM  Smoking habit  Drinking habit  History of complications  \
0              72.0            0.0             0.0                       1.0   
1              36.0            2.0             0.0                       1.0   
2             168.0            2.0             0.0                       1.0   
3             144.0            2.0             1.0                       1.0   
4              72.0            1.0             0.0                       1.0   

   History of hypertension  History of dyslipidemia  ...  Urinary sugar at 3M  \
0                      1.0                      1.0  ...                  NaN   
1                      0.0                      1.0  ...                  4.0   
2                      0.0                      1.0  ...                  4.0   
3                      1.0                      1.0  ...                  0.0   
4                      1.0                      1.0  ...                  0.0   

   Urinary sugar at 12M  Urinary albumin/creatinine ratio at 0M   \
0                   0.0                                     53.3   
1                   NaN                                      NaN   
2                   NaN                                      NaN   
3                   0.0                                      NaN   
4                   0.0                                     30.0   

   Urinary albumin/creatinine ratio at 3M   \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                     27.1   

   Urinary albumin/creatinine ratio at 12M   eGFR at 0M  eGFR at 1M  \
0                                       6.4   55.940643   52.670513   
1                                       NaN         NaN         NaN   
2                                       NaN         NaN         NaN   
3                                       NaN         NaN         NaN   
4                                      56.6   77.754106   87.433453   

   eGFR at 3M  eGFR at 12M  Risk of Nephropathy  
0   50.686818    61.643003                  0.0  
1         NaN          NaN                  0.0  
2         NaN          NaN                  0.0  
3         NaN          NaN                  0.0  
4   89.274571    83.961202                  1.0  

[5 rows x 165 columns]

In [4]:
print(neph_df.info())  #基本信息
round(neph_df.describe()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 779 entries, 0 to 778
Columns: 165 entries, Age to Risk of Nephropathy
dtypes: float64(151), object(14)
memory usage: 1004.3+ KB
None


Age    Sex  Body height  at 0M   Waist circumstances at 0M   \
count  778.0  778.0                765.0                       467.0   
mean    64.0    1.0                163.0                        88.0   
std     12.0    0.0                 10.0                        11.0   
min     27.0    1.0                135.0                        38.0   
25%     56.0    1.0                155.0                        82.0   
50%     64.0    1.0                164.0                        87.0   
75%     73.0    2.0                170.0                        94.0   
max     98.0    2.0                188.0                       187.0   

       Duration of T2DM  Smoking habit  Drinking habit  \
count             713.0          748.0           745.0   
mean              106.0            1.0             0.0   
std                81.0            1.0             1.0   
min                 1.0            0.0             0.0   
25%                48.0            0.0             0.0   
50%                90.0            0.0             0.0   
75%               141.0            2.0             1.0   
max               480.0            2.0             1.0   

       History of complications  History of hypertension  \
count                     778.0                    778.0   
mean                        1.0                      1.0   
std                         0.0                      0.0   
min                         0.0                      0.0   
25%                         1.0                      0.0   
50%                         1.0                      1.0   
75%                         1.0                      1.0   
max                         1.0                      1.0   

       History of dyslipidemia  ...  Urinary sugar at 3M  \
count                    778.0  ...                665.0   
mean                       1.0  ...                  1.0   
std                        0.0  ...                  1.0   
min                        0.0  ...                  0.0   
25%                        0.0  ...                  0.0   
50%                        1.0  ...                  0.0   
75%                        1.0  ...                  1.0   
max                        1.0  ...                  5.0   

       Urinary sugar at 12M  Urinary albumin/creatinine ratio at 0M   \
count                 623.0                                    360.0   
mean                    1.0                                     67.0   
std                     1.0                                    271.0   
min                     0.0                                      0.0   
25%                     0.0                                      6.0   
50%                     0.0                                     13.0   
75%                     1.0                                     31.0   
max                     5.0                                   3451.0   

       Urinary albumin/creatinine ratio at 3M   \
count                                    291.0   
mean                                      69.0   
std                                      261.0   
min                                        0.0   
25%                                        6.0   
50%                                       12.0   
75%                                       32.0   
max                                     3525.0   

       Urinary albumin/creatinine ratio at 12M   eGFR at 0M  eGFR at 1M  \
count                                     358.0       690.0       643.0   
mean                                       88.0        79.0        76.0   
std                                       356.0        20.0        20.0   
min                                         0.0        17.0        16.0   
25%                                         6.0        66.0        64.0   
50%                                        12.0        77.0        74.0   
75%                                        30.0        91.0        89.0   
max                                      3678.0       1

In [15]:
#缺失值观察
print(neph_df.isnull().any()) #判断哪些列包含缺失值

#显示各特征缺失比例图像
plt.style.use('ggplot')     #设置背景网格
plt.rcParams['font.family']='Microsoft YaHei'   #字体

null_percentage = neph_df.isnull().sum()/len(neph_df)#计算每列缺失比例
# print ('The null data percentage is:',null_percentage)
null_percentage = null_percentage.reset_index()
null_percentage.columns = ['column_name','column_value']
ind = np.arange(null_percentage.shape[0])
fig , ax = plt.subplots(figsize = (16, 30))
rects = ax.barh(ind,null_percentage.column_value.values,color='red')
ax.set_yticks(ind)
ax.set_yticklabels(null_percentage.column_name.values,rotation='horizontal')
ax.set_xlabel("各基本特征缺失数据值")
plt.show()

Age                                         True
Sex                                         True
Body height  at 0M                          True
Waist circumstances at 0M                   True
Duration of T2DM                            True
Smoking habit                               True
Drinking habit                              True
History of complications                    True
History of hypertension                     True
History of dyslipidemia                     True
History of hyperuricemia                    True
History of retinopathy                      True
History of arteriosclerosis obliterans      True
History of atrial fibrillation              True
History of kidney disease                   True
History of liver disease                    True
History of myocardial infarction            True
History of cerebral infarction              True
History of angina pectoris                  True
History of heart failure                    True
Systolic blood press

<Figure size 1152x2160 with 1 Axes>

**分析结果**：由上面的基本分析，该数据包含779个样本，165列，其中最后一列Risk of Nephropathy为label。由label的值可知该问题**是个二分类问题**。
而且数据集中存在大量缺失值。下面对数据进行处理

### Step 2：数据预处理

1. 缺失值处理：对缺少标签的样本删除，对缺失值太多的特征进行删除，对剩下的缺失值进行填充
1. 特征分析，选择



In [19]:
# 删除缺少标签的行
neph_data = neph_df.dropna(subset=['Risk of Nephropathy'])
neph_data.shape

(778, 165)

In [5]:
import seaborn as sns

# 查看数据类别是否平衡
fig, ax = plt.subplots()
sns.set()
ax = sns.distplot(neph_data['Risk of Nephropathy'],kde=False)

print(neph_data['Risk of Nephropathy'].value_counts())

0.0    556
1.0    222
Name: Risk of Nephropathy, dtype: int64


<Figure size 432x288 with 1 Axes>

In [20]:
# 保留缺失值比例不大于0.2的特征
feats = []
null_percentage = neph_data.isnull().sum()/len(neph_data)#计算每列缺失比例

for i in range(0,null_percentage.shape[0]):
    if null_percentage.values[i] <= 0.3:
        feats.append(null_percentage.index[i])

print(feats, len(feats))

['Age', 'Sex', 'Body height  at 0M ', 'Duration of T2DM', 'Smoking habit', 'Drinking habit', 'History of complications', 'History of hypertension', 'History of dyslipidemia', 'History of hyperuricemia', 'History of retinopathy', 'History of arteriosclerosis obliterans ', 'History of atrial fibrillation', 'History of kidney disease', 'History of liver disease', 'History of myocardial infarction', 'History of cerebral infarction', 'History of angina pectoris', 'History of heart failure', 'Systolic blood pressure at 0M ', 'Systolic blood pressure at 1M ', 'Systolic blood pressure at 3M ', 'Systolic blood pressure at 12M ', 'Diastolic blood pressure at 0M', 'Diastolic blood pressure at 1M', 'Diastolic blood pressure at 3M', 'Diastolic blood pressure at 12M', 'Pulse rate at 0M', 'Pulse rate at 1M', 'Pulse rate at 3M', 'Body weight at 0M', 'Body weight at 1M', 'Body weight at 3M', 'Body weight at 12M', 'BMI at 0M', 'BMI at 1M', 'BMI at 3M', 'BMI at 12M', 'HbA1c (NGSP*) at 0M', 'HbA1c (NGSP*)

In [21]:
neph_data = neph_data[feats]
print(neph_data.shape)

(778, 124)


In [22]:
# 对剩下的缺失值进行填充
#离散变量，使用众数填充
feat_a = ['Sex', 'Body height  at 0M ', 'Duration of T2DM', 'Smoking habit', 'Drinking habit', 'History of complications', 'History of hypertension', 
'History of dyslipidemia', 'History of hyperuricemia', 'History of retinopathy', 'History of arteriosclerosis obliterans ', 'History of atrial fibrillation', 
'History of kidney disease', 'History of liver disease', 'History of myocardial infarction', 'History of cerebral infarction', 'History of angina pectoris', 
'History of heart failure']

neph_data[feat_a]=neph_data[feat_a].fillna(method='pad')
# neph_data.fillna(neph_data.mean(),inplace=True)
#使用KNN填充数据 https://mp.weixin.qq.com/s/B-C-E1WtHFnSm-V8MyETvA
# !pip install fancyimpute  #安装相应的包
from fancyimpute import KNN

neph_data = pd.DataFrame(KNN(k=3).fit_transform(neph_data), columns=neph_data.columns)

Imputing row 1/778 with 3 missing, elapsed time: 0.403
Imputing row 101/778 with 69 missing, elapsed time: 0.435
Imputing row 201/778 with 36 missing, elapsed time: 0.456
Imputing row 301/778 with 1 missing, elapsed time: 0.475
Imputing row 401/778 with 14 missing, elapsed time: 0.497
Imputing row 501/778 with 0 missing, elapsed time: 0.509
Imputing row 601/778 with 69 missing, elapsed time: 0.549
Imputing row 701/778 with 83 missing, elapsed time: 0.576


In [59]:
print(neph_data.isnull().any()) #判断哪些列包含缺失值

Age                                        False
Sex                                        False
Body height  at 0M                         False
Duration of T2DM                           False
Smoking habit                              False
Drinking habit                             False
History of complications                   False
History of hypertension                    False
History of dyslipidemia                    False
History of hyperuricemia                   False
History of retinopathy                     False
History of arteriosclerosis obliterans     False
History of atrial fibrillation             False
History of kidney disease                  False
History of liver disease                   False
History of myocardial infarction           False
History of cerebral infarction             False
History of angina pectoris                 False
History of heart failure                   False
Systolic blood pressure at 0M              False
Systolic blood press

In [23]:
# c = neph_data[['Sex','Risk of Nephropathy']].groupby(['Sex'],as_index=False).agg({'Risk of Nephropathy':sum})
# print(c)

neph_data['Drinking habit'].value_counts()

0.0    394
1.0    384
Name: Drinking habit, dtype: int64

In [24]:
#把数据标签和特征划分
from sklearn import preprocessing

#把标签列取出
neph_Y = neph_data['Risk of Nephropathy']
print(neph_Y.head())

#把标签列删除
neph_X = neph_data.drop(columns='Risk of Nephropathy')

# min_max_scaler = preprocessing.MinMaxScaler() 
# neph_X = min_max_scaler.fit_transform(neph_X)
# print(neph_X)

0    0.0
1    0.0
2    0.0
3    0.0
4    1.0
Name: Risk of Nephropathy, dtype: float64


In [38]:
from sklearn.decomposition import PCA   #加载PCA算法包

pca=PCA(n_components=20)  #加载PCA算法，设置降维后主成分数目为10
reduced_x=pca.fit_transform(neph_X)#对样本进行降维
print(reduced_x.shape)

(778, 20)


由于对数据特征表示的意思，以及每个特征的正常取值范围不太了解，仅仅对数据做以上简单处理，在实际科研，或者临床建模中，应对特征进行熟悉的了解（借助医生），可进一步筛选和衍生新的特征（如教案中的“白球比”）

### Step 3: 模型构建：模型选择，训练 


1. 单一模型
1. 模型集成

In [25]:
from sklearn.model_selection import train_test_split

#数据量较少，80%作为训练， 20%作为测试
x_train,x_test,y_train,y_test = train_test_split(neph_X, neph_Y, test_size=0.2)

print(x_train.shape,y_train.shape)
print(x_test.shape,y_test.shape)

(622, 123) (622,)
(156, 123) (156,)


In [26]:
import xgboost as xgboost
from sklearn.ensemble import RandomForestClassifier
import sklearn.svm as svm

#随机森林
forest = RandomForestClassifier(n_estimators= 20, oob_score=True, max_depth= 5)
forest.fit(x_train,y_train)

#SVM
sv_k = svm.SVC(kernel='poly', degree=5, probability=True)
sv_k.fit(x_train,y_train)

#xgboost
# from collections import Counter
# cnt = Counter(neph_Y)
xgb = xgboost.XGBClassifier(n_estimators= 20,  max_depth= 5,gamma=2, eval_metric='auc', min_child_weight=4)
xgb.fit(x_train, y_train)

XGBClassifier(eval_metric='auc', gamma=2, max_depth=5, min_child_weight=4,
              n_estimators=20)

In [27]:
from sklearn.metrics import accuracy_score,precision_score,recall_score,confusion_matrix
from sklearn.metrics import roc_auc_score
#测试随机森林模型
y_pred_rt = forest.predict(x_test)

print("随机森林 Acc: " ,accuracy_score(y_pred_rt, y_test))

print(confusion_matrix(y_pred_rt, y_test))

#测试随机森林模型
y_pred_svm = sv_k.predict(x_test)

print("SVM Acc: " ,accuracy_score(y_pred_svm, y_test))

print(confusion_matrix(y_pred_svm, y_test))

#测试随机森林模型
y_pred_xgb = xgb.predict(x_test)

print("xgboost Acc: " ,accuracy_score(y_pred_xgb, y_test))

print(confusion_matrix(y_pred_xgb, y_test))

随机森林 Acc:  0.7628205128205128
[[118  37]
 [  0   1]]
SVM Acc:  0.75
[[117  38]
 [  1   0]]
xgboost Acc:  0.7371794871794872
[[114  37]
 [  4   1]]


In [28]:
#绘制ROC曲线
from sklearn.metrics import roc_curve
y_pred_p = forest.predict_proba(x_test)
y_pred_p2 = sv_k.predict_proba(x_test)
y_pred_p3 = xgb.predict_proba(x_test)

fpr, tpr, thresholds = roc_curve(y_test, y_pred_p[:,1], pos_label=1)
fpr2, tpr2, thresholds2 = roc_curve(y_test,y_pred_p2[:,1], pos_label=1)
fpr3, tpr3, thresholds3 = roc_curve(y_test,y_pred_p3[:,1], pos_label=1)

AUC_ROC = roc_auc_score(y_test, y_pred_p[:,1])
AUC_ROC2 = roc_auc_score(y_test, y_pred_p2[:,1])
AUC_ROC3 = roc_auc_score(y_test, y_pred_p3[:,1])

roc_curve =plt.figure()
plt.plot(fpr,tpr,'-',label='RandomForest (AUC = %0.4f)' % AUC_ROC)
plt.plot(fpr2,tpr2,'-',label='SVM (AUC = %0.4f)' % AUC_ROC2)
plt.plot(fpr3,tpr3,'-',label='Xgboost (AUC = %0.4f)' % AUC_ROC3)

plt.title('ROC curve')
plt.xlabel("FPR (False Positive Rate)")
plt.ylabel("TPR (True Positive Rate)")
plt.legend(loc="lower right")

<Figure size 432x288 with 1 Axes>

**特征选择**：训练好的模型中，根据特征的重要性（贡献度）重新选择一组特征，再训练模型。

In [29]:
from sklearn.feature_selection import SelectFromModel

# select features using threshold
selection = SelectFromModel(xgb, threshold=0.000001, prefit=True)
select_X_train = selection.transform(x_train)
print(select_X_train.shape)
# train model
selection_model =  xgboost.XGBClassifier(n_estimators= 10, gamma=0.2)
selection_model.fit(select_X_train, y_train)
# eval model
select_X_test = selection.transform(x_test)
select_y_pred_xgb = selection_model.predict(select_X_test)

print("after select xgboost Acc: " ,accuracy_score(select_y_pred_xgb, y_test))

print(confusion_matrix(select_y_pred_xgb, y_test))

(622, 81)
after select xgboost Acc:  0.7628205128205128
[[118  37]
 [  0   1]]


单纯从准确率这个指标来看，通过选择更重要几个特征训练模型可以有一定提升。不过通过观察混淆矩阵，其实相差不大。

下面采用模型组合的方式进行训练
**模型组合**：对上面几种模型，用Voting的方法进行集成，采用软投票（即概率）得到分类结果。

In [33]:
from sklearn.ensemble import VotingClassifier

estimators = []
#随机森林
model1 = RandomForestClassifier(n_estimators= 20) #设置类别平衡
estimators.append(('rt', model1))

#SVM
model2 = svm.SVC(kernel='poly', degree=3, probability=True) #设置类别平衡
estimators.append(('svm', model2))

#xgboost
model3 = xgboost.XGBClassifier(n_estimators= 20, gamma=0.2) #设置类别平衡
estimators.append(('xgboost', model3))

ensemble = VotingClassifier(estimators)

ensemble.fit(x_train,y_train)

VotingClassifier(estimators=[('rt', RandomForestClassifier(n_estimators=20)),
                             ('svm', SVC(kernel='poly', probability=True)),
                             ('xgboost',
                              XGBClassifier(gamma=0.2, n_estimators=20))])

In [34]:
#测试模型
score = ensemble.score(x_test, y_test)
print(score)

0.7628205128205128


### Step4:  结果分析

经过多次调参，模型还是不能比较好的拟合数据，如何比较好的提升模型精度：个人觉得，对于数据的缺失值填充，特征的选择和各特征的含义再深入了解，把数据质量变高，模型有比较大可能提升！